### DISCUSSION TITLE
Ability to set ssl/tls for smtp in mail settings with environment variable

### DISCUSSION THREAD

**Discussion by EpocSquadron** [craftcms/cms#8989] [Ideas] [answered]
### Description

We can currently set the smtp credentials, server and port all with environment variables. However, it appears the dropdown to select ssl vs tls for the connection is not able to be set to an environment variable. We are currently using mailtrap.io for testing environments and a proper transactional email service in production (although not one of the common ones). Mailtrap requires the connection be over tls (mail settings test fails when set to ssl), while production requires ssl. This wouldn't be quite such a problem except that we are enforcing project config (and have admin changes off in all non-local environments), and this setting is tracked in that.  

I'm hesitant to override the transport in config/app.php as in #5975 as it seems kind of invasive, especially seeing as all of the other fields in the smtp settings are configurable with environment variables. Is it possible to add support for configuring this in a way that project config is compatible with?

**Comment by nordentwickler** [comment]
Completely agree with this. We are using laradock with maildev for our dev environment and the most small websites still user smtp with tls for sending emails. Maildev breaks with tls setting on - controlling this via environment variables would be great.

**Comment by Mosnar** [comment]
Related: #6403

**Comment by ryanirelan** [comment]
This takes on new meaning with Nitro's support for MailHog. MailHog doesn't support authentication nor encryption but those cannot be set via environment variables. 

[See docs:](https://craftcms.com/docs/nitro/2.x/services/mailhog.html)
> Transport Type: SMTP
> Host Name: mailhog.service.nitro
> Port: 1025
> Use Authentication: Unchecked (default)
> Encryption Method: None (default)
> Timeout: 10 (default)

**Comment by Antimated** [comment]
This would be a very nice addition indeed. We use mailhog locally which does not support authentication nor encryption (lika @ryanirelan) stated. For our staging and production environments we use postmark which requires you to use TLS encryption. Very annoying that we constantly have to switch between no encryption and TLS when we are testing locally vs staging.

**Comment by peteeveleigh** [comment]
Whilst I agree you should be able to set ALL mail settings via environment vars there is another way to get around this.
You can configure the mail transport set up in config/app.php.
This can then use your environment variables to populate any of the values.

In the components section of your app.php file you can do something like

```
'mailer' => function() {
            // Get the stored email settings
            $settings = App::mailSettings();

            // Override the transport adapter class
            $settings->transportType = PostmarkAdapter::class;

            // Override the transport adapter settings
            $settings->transportSettings = [
                'token' => getenv('POSTMARK_API_TOKEN'),
                'messageStream' => getenv('POSTMARK_STREAM_ID') ?? ''
            ];

            // Create a Mailer component config with these settings
            $config = App::mailerConfig($settings);
            return Craft::createObject($config);
        }
```

**Comment by brandonkelly** [comment, marked as answer]
This is now resolved for the next Craft 3 release 🚀 

See details at #10111

**Reply by brandonkelly** [reply]
Craft 3.7.22 is out now with those changes. And I managed to get the “Use authentication” setting covered as well (615ea4044a5f1378f0123209abec1068c91379da).

**Reply by mikewink** [reply]
Hello Brandon, 

I have not figured yet out how to name the .env variable so that it shows up for Encryption Method. The dropdown shows None, SSL and TLS. I can't see other suggestions at all nor can I enter my own variable name. It just defaults back to none then. What am I missing? 🤔😃 Thank you in advance.

I tried: 

```
EMAIL_ENCRYPTION_METHOD=SSL
ENCRYPTION_METHOD=SSL
ENCRYPTIONMETHOD=SSL
ENCRYPTION=SSL
```

<img width="430" alt="Screenshot 2021-11-24 at 22 54 48" src="https://user-images.githubusercontent.com/85063626/143317599-21605023-6123-4f03-b78b-110b6f2de757.png">

**Reply by michaelij** [reply]
Hi Brandon,

Also having issues selecting any environment variables in the Encryption Method dropdown in particular. The dropdown seems to show only the three preset options, manually typing the env variable does not work either.

**Reply by brandonkelly** [reply]
It will only show environment variables that match the available option values (`none`, `ssl`, or `tls`). Should probably make that clearer…

**Reply by mikewink** [reply]
Works Brandon! Thank you very much for clarifying. 🏆

**Reply by brandonkelly** [reply]
Just released 3.7.23 which should make this more clear. There is now a tip below the field that says:

> This can be set to an environment variable with a value of one of the selectable option values.

And the dropdown now shows each of the native option values beside their labels:

<img width="663" alt="A screenshot of the Encryption Method setting in 3.7.23, expanded to reveal all of its options. Each option shows both the option label and its value." src="https://user-images.githubusercontent.com/47792/143621185-0328a355-3053-4385-9ff7-e33b740f1d9a.png">

The “Use authentication” setting also shows a tip with the supported environment varialeble values:

<img width="676" alt="A screenshot of the “Use authentication” setting in 3.7.23, listing the supported environment variable values: yes, no, on, off, true, false, 0, and 1." src="https://user-images.githubusercontent.com/47792/143621349-15e66383-6a96-4db4-a08a-55b866c26325.png">

**Reply by darylknight** [reply]
I just came here to post a bug and found this discussion too. I had an env variable set to "SSL" and it didn't show up in the list, so it just looked like "Use Authentication" was bugged and not looking for ENV variables. Now that I've changed the value to "ssl", it shows up in the list.
